core/src/main/java/io/parsingdata/metal/data/ImmutableList.java

```diff
@@ -100,4 +102,27 @@ public int hashCode() {
         return Objects.hash(getClass(), head, tail);
     }
 
+    @Override
+    public Iterator<T> iterator() {
+        final AtomicReference<ImmutableList<T>> next = new AtomicReference<>(this);
+        return new Iterator<T>() {
+
+            @Override
+            public boolean hasNext() {
+                return !next.get().isEmpty();
+            }
+
+            @Override
+            public T next() {
+                final ImmutableList<T> list = next.get();
+                if (list.isEmpty()) {
+                    return null;
```

Comment:
Or throw an exception, e.g. `NoSuchElementException`, since there is a `hasNext`method?